### PR TITLE
Changed label application rule to support multiple labels and entity-types and single dynamic label

### PR DESCRIPTION
**Pull request by kamaleshnneerasa**
## Description
Please include a summary of the change, motivation and context.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto [hunk 1]
```diff
@@ -70,18 +70,23 @@ message LabelApplicationRuleData {
...
 
   message Action {
-    string entity_type = 1; // captures the entity type - "API", "Service"
+    repeated string entity_type = 1; // captures the entity type - "API", "Service"
     Operation operation = 2;
     oneof value {
-      string static_label_id = 3;
+      StaticLabels static_labels = 3;
       DynamicLabel dynamic_label = 4;
+      string label_replacement_token = 5;
```
Comment (saxenakshitiz) [inline label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto:78]: dynamic_label_key

Comment (kamaleshnneerasa) [reply label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto:78]: done.

File: label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto [hunk 2]
```diff
@@ -70,18 +70,23 @@ message LabelApplicationRuleData {
...
     }
 
     enum Operation {
       OPERATION_UNSPECIFIED = 0;
       OPERATION_MERGE = 1;
     }
 
+    message StaticLabels{
+      repeated string static_label_id = 1;
```
[the remark this answers is not in the document]
Comment (kamaleshnneerasa) [reply label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto:87]: done.

File: label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto [hunk 3]
```diff
@@ -70,18 +70,23 @@ message LabelApplicationRuleData {
...
   message Action {
-    string entity_type = 1; // captures the entity type - "API", "Service"
+    repeated string entity_types = 1; // captures the entity type - "API", "Service"
     Operation operation = 2;
     oneof value {
-      string static_label_id = 3;
-      DynamicLabel dynamic_label = 4;
+      StaticLabels static_labels = 3;
+      string dynamic_label_key = 4;
+      DynamicLabel dynamic_label = 5;
```
Comment (aaron-steinfeld) [inline label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto:78]: nit dynamic_label_expression perhaps? To disambiguate between the simpler form on 77.

File: label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto [hunk 4]
```diff
@@ -70,18 +70,23 @@ message LabelApplicationRuleData {
...
     }
 
     enum Operation {
       OPERATION_UNSPECIFIED = 0;
       OPERATION_MERGE = 1;
     }
 
+    message StaticLabels{
```
Comment (aaron-steinfeld) [inline label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto:86]: missing space here - make sure to run formatter.

File: label-application-rule-config-service-impl/src/test/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImplTest.java [hunk 5]
```diff
@@ -261,9 +261,11 @@ private LabelApplicationRule createCompositeRule() {
 
   private LabelApplicationRuleData.Action buildAction() {
     return LabelApplicationRuleData.Action.newBuilder()
-        .setEntityType("API")
+        .addAllEntityTypes(List.of("API"))
```
Comment (aaron-steinfeld) [inline label-application-rule-config-service-impl/src/test/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImplTest.java:264]: nit: addEntityType (repeated types take a single or collection, so no need to wrap to match the collection api)

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java [hunk 6]
```diff
@@ -127,21 +129,43 @@ private void validateUnaryCondition(UnaryCondition unaryCondition) {
...
+        break;
       default:
         throwInvalidArgumentException(
             String.format("Unexpected Case in %s:%n %s", getName(action), printMessage(action)));
     }
   }
 
+  void validateEntityTypes(List<String> entityTypes) {
+    Set<String> entityTypesSet = new HashSet<>(entityTypes);
+    if (entityTypesSet.size() < entityTypes.size()) {
```
Comment (aaron-steinfeld) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java:153]: minor nit: One less variable name and one less potential bug (that is, it's easy to flip the >, while != is symmetric and gives us the same info)
```suggestion
    if (Set.copyOf(entityTypes).size() != entityTypes.size()) {
```

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java [hunk 7]
```diff
@@ -127,21 +129,43 @@ private void validateUnaryCondition(UnaryCondition unaryCondition) {
...
+      throwInvalidArgumentException(String.format("Duplicate entity types %s", entityTypes));
+    }
+  }
+
+  void validateStaticLabels(Action.StaticLabels staticLabels) {
+    if (staticLabels.getIdsCount() == 0) {
```
Comment (aaron-steinfeld) [inline label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java:159]: this check is handled by `validateNonDefaultPresenceOrThrow`

**Comment by aaron-steinfeld** [comment]
> These changes will obviously break protos verification. How should we go about fixing it? CC:: @aaron-steinfeld

We can override since they're not in use yet.

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java (hunk 6 above)
Comment (kamaleshnneerasa) [reply label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java:153]: done.

File: label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java [hunk 8]
```diff
@@ -173,6 +189,12 @@ public void validateLabelExpression(Action.DynamicLabel dynamicLabel) {
     }
   }
 
+  private void validateNoDuplicates(List<String> list, String errorMessage) {
```
[the remark this answers is not in the document]
Comment (kamaleshnneerasa) [reply label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleValidatorImpl.java:192]: What is the advantage of sending errorMessage and msgArgs separately?

**Pull request merged**